### ISSUE TITLE
binary stubs for windows leak "cd" output into $dir

### ISSUE THREAD

**Issue by ktomk** [composer/composer#6567]
on windows 10 with git bash using the binary proxy stubs (sh shell) caused me issues as it reported file not found (as the proxy is not working)

the actual file to resolve to (or more correct: it's directory from which it was going to be invoked) - which looked correct - was somewhat prefixed with another directory and a space.

when looking into the issue it became apparent to me that the line which resolved the current directory:

    dir=$(d=\${0%[/\\\\]*}; cd "\$d"; cd $binDir && pwd)
                            ^^^^^^^^

was putting that directory in from the first `cd` command in front of the output of `pwd`.

**Comment by ktomk** [comment]
solution in my case is to mute the output by redirecting it's standard output to `/dev/null`. i'll file a patch via a pull request.

**Comment by TKCk24** [comment]
can confirm this behavior, point in case is if the project is in a directory within bash's CDPATH, the *cd* command will output the path cd'ed into.

**Comment by ktomk** [comment]
For reference on cd output with CDPATH: https://unix.stackexchange.com/q/245453/9340

**Comment by johnstevenson** [comment]
@alcohol composer-setup does [this](https://github.com/composer/windows-setup/blob/master/src/shims/composer#L3)

`dir=$(cd "${0%/*}" && pwd)`

Is this ok?

**Comment by TKCk24** [comment]
@johnstevenson: Looks like the cd could leak to stdout as well.

**Comment by johnstevenson** [comment]
@TKCk24 Thanks. Could you give me an example to test with please.

**Comment by TKCk24** [comment]
@johnstevenson: Please see this test-example, is that what you're asking for?

```
#!/bin/bash

CDPATH="$(pwd)"

# show CDPATH
echo "CDPATH: ${CDPATH}"

# when found via CDPATH, echoes
cd ""
echo "- ^^ output ^^ -"

# when found directly, no echo
cd "."
echo "- ^^ no-output ^^ -"

```

**Comment by johnstevenson** [comment]
@TKCk24 Thanks for that. However, if I change `cd ""` to `dir=$(cd "")` to match the usage, then there is no output. So I am presuming that this is okay ??

**Comment by TKCk24** [comment]
@johnstevenson: No, the output goes into the variable:

~~~
#!/bin/bash

CDPATH="$(pwd)"

# show CDPATH
echo "CDPATH: ${CDPATH}"

# when found via CDPATH, echoes
dir="$(cd ""; pwd)";
echo "${dir}"
# you will see two lines, one from the cd, one from the pwd
~~~

Output:

~~~
CDPATH: /c/Users/tom.klingenberg/proj/tests
/c/Users/user/proj/tests
/c/Users/user/proj/tests
~~~

**Comment by johnstevenson** [comment]
@TKCk24: Ah, I am an idiot. Thanks. I'll fix this in composer-setup.

The fix for this issue should presumably handle all cds, so:

`dir=$(d=\${0%[/\\\\]*}; cd "\$d" > /dev/null; cd $binDir > /dev/null; pwd)`

**Comment by ktomk** [comment]
@johnstevenson: Not all `cd`s, the `cd $binDir` is with a relative path so it is not affected.

**Comment by johnstevenson** [comment]
@TKCk24: Okay, I hadn't realized that was always the case for `$binDir`. Thanks for your help.

**Comment by ktomk** [comment]
Sure, I also have not quadruple checked it, but when I look into the script created, it's one up  `..` into `vendor` and then to the packages bin-script, e.g. `../phpunit/phpunit/phpunit` for example for PHP Unit. I'd say this is always the case for any other bin-script as well as they come from inside vendor. The `vendor/bin` directory is for offering a common location to access these.

If you want to reproduce yourself on a Windows system with Git-Bash, just put the projects directory (or one up I think I have in that case) into the CDPATH. Then calling from the root directory of that project the (sh-shell) proxy script triggers that, e.g. `vendor/bin/phpunit`. Using `./vendor/bin/phpunit` does not trigger it (as it is a relative path then that resolves already against PWD).

**Comment by johnstevenson** [comment]
Yup, I can reproduce that. Given a project path of `path/to/project`:
```
> cd path/to/project 
> export CDPATH="$(pwd)"
> vendor/bin/phpunit
```
`$d` is set to `vendor/bin` and so `cd "$d"` finds this in CDPATH and results in `dir` being set to:

```
path/to/project/vendor/bin
path/to/project/vendor/phpunit/phpunit
```
rather than `path/to/project/vendor/phpunit/phpunit`

**Comment by Seldaek** [comment]
Closed by #6568
